### ISSUE TITLE
Panic when verifying token with store if key is missing 

### ISSUE THREAD

**Issue by fschuetz** [mikkyang/rust-jwt#61]
Hi

When trying to verify a token that was signed by a key that is not present in the store a panic occurs. The store is using a BTreeMap<String, Hmac<Sha384>> as data structure. I think I found the error at store.rs:17 in the implementation of fn get(&self, key_id: &str) -> Option<&A>. Instead of properly checking if the key is there, it assumes so and returns Some(&self[key_id]). The BTreeMap implementation does panic when a key is not found when using btree[key] syntax.

Full function is:

`
fn get(&self, key_id: &str) -> Option<&A> {
        Some(&self[key_id])
    }
`

I assume  a better way to do it would be to check key existence with .contains_key().

Stack trace:
`
---- tests::verifying_token_failure_missing_key_in_store stdout ----
thread 'tests::verifying_token_failure_missing_key_in_store' panicked at 'no entry found for key', /Users/fschuetz/.cargo/registry/src/github.com-1ecc6299db9ec823/jwt-0.11.0/src/algorithm/store.rs:17:15
stack backtrace:
   0:        0x10fdac484 - std::backtrace_rs::backtrace::libunwind::trace::hf8598643f9eebc26
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/std/src/../../backtrace/src/backtrace/libunwind.rs:96
   1:        0x10fdac484 - std::backtrace_rs::backtrace::trace_unsynchronized::h19a3bd07f8cb7f35
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/std/src/../../backtrace/src/backtrace/mod.rs:66
   2:        0x10fdac484 - std::sys_common::backtrace::_print_fmt::h3fc167841ad5be48
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/std/src/sys_common/backtrace.rs:79
   3:        0x10fdac484 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hbb144aff6cb2709e
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/std/src/sys_common/backtrace.rs:58
   4:        0x10fdcc2e0 - core::fmt::write::h0ce880d33cd2a300
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/core/src/fmt/mod.rs:1080
   5:        0x10fd14029 - std::io::Write::write_fmt::h6670478f92af613f
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/std/src/io/mod.rs:1517
   6:        0x10fda6b0c - std::io::impls::<impl std::io::Write for alloc::boxed::Box<W>>::write_fmt::hd522689453af032f
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/std/src/io/impls.rs:179
   7:        0x10fdae1a5 - std::sys_common::backtrace::_print::h1418657e0e107c4e
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/std/src/sys_common/backtrace.rs:61
   8:        0x10fdae1a5 - std::sys_common::backtrace::print::h3c3eba71f67c7cbd
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/std/src/sys_common/backtrace.rs:48
   9:        0x10fdae1a5 - std::panicking::default_hook::{{closure}}::hb543bf3b05c8104b
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/std/src/panicking.rs:208
  10:        0x10fdade2e - std::panicking::default_hook::h7015a8c764df2477
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/std/src/panicking.rs:224
  11:        0x10fdae75b - std::panicking::rust_panic_with_hook::hb4c39ab085f5a3ba
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/std/src/panicking.rs:577
  12:        0x10fdae2db - std::panicking::begin_panic_handler::{{closure}}::h85978c5141656d56
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/std/src/panicking.rs:484
  13:        0x10fdac8f8 - std::sys_common::backtrace::__rust_end_short_backtrace::hba09eb580b3852d8
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/std/src/sys_common/backtrace.rs:153
  14:        0x10fdae29a - rust_begin_unwind
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/std/src/panicking.rs:483
  15:        0x10fdd576f - core::panicking::panic_fmt::h33aee6f32da9c3a4
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/core/src/panicking.rs:85
  16:        0x10fdd55fa - core::option::expect_failed::he41d9b6b925c99b1
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/core/src/option.rs:1213
  17:        0x10fce8144 - core::option::Option<T>::expect::h518707def542c752
                               at /Users/fschuetz/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/library/core/src/option.rs:333
  18:        0x10fd0de7d - <alloc::collections::btree::map::BTreeMap<K,V> as core::ops::index::Index<&Q>>::index::hfd64eec3b7a1e475
                               at /Users/fschuetz/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/library/alloc/src/collections/btree/map.rs:2080
  19:        0x10fd0ea93 - <T as jwt::algorithm::store::Store>::get::hddd30c68309ee59e
                               at /Users/fschuetz/.cargo/registry/src/github.com-1ecc6299db9ec823/jwt-0.11.0/src/algorithm/store.rs:17
  20:        0x10fce2337 - <jwt::Token<H,C,jwt::token::Unverified> as jwt::token::verified::VerifyWithStore<jwt::Token<H,C,jwt::token::Verified>>>::verify_with_store::hcb29a5f699bcb29f
                               at /Users/fschuetz/.cargo/registry/src/github.com-1ecc6299db9ec823/jwt-0.11.0/src/token/verified.rs:59
  21:        0x10fce8e34 - <&str as jwt::token::verified::VerifyWithStore<jwt::Token<H,C,jwt::token::Verified>>>::verify_with_store::h255f27a2cf6bc617
                               at /Users/fschuetz/.cargo/registry/src/github.com-1ecc6299db9ec823/jwt-0.11.0/src/token/verified.rs:92
  22:        0x10fcdb2e7 - auth::verify_token_with_store::h996870a2b23aa8d9
                               at /Users/fschuetz/Development/smart_irrigation/irrigation_sensor_actor_node/lib/auth/src/lib.rs:678
  23:        0x10fcea385 - auth::tests::verifying_token_failure_missing_key_in_store::h762ecdc17a252b8f
                               at /Users/fschuetz/Development/smart_irrigation/irrigation_sensor_actor_node/lib/auth/src/lib.rs:835
  24:        0x10fce3351 - auth::tests::verifying_token_failure_missing_key_in_store::{{closure}}::he2ec59e7a53d560f
                               at /Users/fschuetz/Development/smart_irrigation/irrigation_sensor_actor_node/lib/auth/src/lib.rs:827
  25:        0x10fcf4201 - core::ops::function::FnOnce::call_once::he61c0107fd7751c3
                               at /Users/fschuetz/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/library/core/src/ops/function.rs:227
  26:        0x10fd3adda - core::ops::function::FnOnce::call_once::h0019030a37b5f8e2
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/core/src/ops/function.rs:227
  27:        0x10fd3adda - test::__rust_begin_short_backtrace::h585ac6549c4e9ffc
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/test/src/lib.rs:517
  28:        0x10fd395ee - <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::hf6fe10f95bab47a9
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/alloc/src/boxed.rs:1042
  29:        0x10fd395ee - <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::hc1191b9e91734212
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/std/src/panic.rs:308
  30:        0x10fd395ee - std::panicking::try::do_call::h481b7bf7e4a11df1
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/std/src/panicking.rs:381
  31:        0x10fd395ee - std::panicking::try::hd1e107dde0012eb0
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/std/src/panicking.rs:345
  32:        0x10fd395ee - std::panic::catch_unwind::hf93bb20c0250cdab
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/std/src/panic.rs:382
  33:        0x10fd395ee - test::run_test_in_process::h92fda288fb96bac7
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/test/src/lib.rs:544
  34:        0x10fd395ee - test::run_test::run_test_inner::{{closure}}::h2b34441b0c20bbc1
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/test/src/lib.rs:450
  35:        0x10fd1349b - std::sys_common::backtrace::__rust_begin_short_backtrace::he8660dbf08444d3b
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/std/src/sys_common/backtrace.rs:137
  36:        0x10fd187b5 - std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}::h504a86aa1475381e
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/std/src/thread/mod.rs:461
  37:        0x10fd187b5 - <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::hbab29a65098269ce
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/std/src/panic.rs:308
  38:        0x10fd187b5 - std::panicking::try::do_call::h472740357bdd006c
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/std/src/panicking.rs:381
  39:        0x10fd187b5 - std::panicking::try::h15000d283be751f4
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/std/src/panicking.rs:345
  40:        0x10fd187b5 - std::panic::catch_unwind::h934d0d641bb43f5f
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/std/src/panic.rs:382
  41:        0x10fd187b5 - std::thread::Builder::spawn_unchecked::{{closure}}::h90d7d0da89b4018c
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/std/src/thread/mod.rs:460
  42:        0x10fd187b5 - core::ops::function::FnOnce::call_once{{vtable.shim}}::h55959f91609d768b
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/core/src/ops/function.rs:227
  43:        0x10fdb4add - <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::h034e756f73043ffd
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/alloc/src/boxed.rs:1042
  44:        0x10fdb4add - <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::hdeb0aa4bdb685311
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/alloc/src/boxed.rs:1042
  45:        0x10fdb4add - std::sys::unix::thread::Thread::new::thread_start::h9514c02363ad7497
                               at /rustc/f68e08933d8f519a9655934fedebbc509661b219/library/std/src/sys/unix/thread.rs:87
  46:     0x7fff6908f109 - __pthread_start
`

Thanks for verifying.

**Comment by mikkyang** [comment]
Hi @fschuetz, thanks for bringing this up.

I've fixed this in #62 for the most common types of `Store` that will be used, `BTreeMap` and `HashMap`, however this does mean that other types will need to implement `Store` themselves. I initially relied on the `Index` trait to automatically implement this for common types, but you're right that stability is more important than saved code, and since `Index` doesn't support checked get, I implemented the `Store` trait for each specific type.
